### ISSUE TITLE
Catalog Apps - how to handle arrays in catalog questions

### ISSUE THREAD

**Issue by lmangani** [rancher/rancher#13952]
Rancher versions: 2.0 master

Apologies in advance if this is documented somewhere and I missed it, but I can't quite figure how to use catalog `questions.yaml` to allow forming `arrays` with arbitrary values dynamically from the Rancher UI, ie:
```
servers:
  - server1
  - server2
```

I imagine this exists as used by elements such Ingress rules, etc. Is there any way or example catalog or questions codeblock I could reference for such behavior?

Thanks in advance!

**Comment by selmison** [comment]
I'm facing the same issue when deploy traefik by means of helm

**Comment by StrongMonkey** [comment]
FYI https://github.com/kubernetes/helm/blob/7cbed3f77bd771dfb46c63733c497e8c05225422/docs/using_helm.md#the-format-and-limitations-of---set. So the correct format would be `key={a, b, c}`

**Comment by selmison** [comment]
key={a, b, c} means that **array/list**:
```
key:
  - a
  - b
```

I need **dict/hash**:
```
 key:
  a: value1
  b: value2
```
The  [`link`](https://github.com/kubernetes/helm/blob/7cbed3f77bd771dfb46c63733c497e8c05225422/docs/using_helm.md#the-format-and-limitations-of---set.) above helped me see the correct syntax for the dict/hash is:

```
 key.a=value1
 key.b=value2
```

Thanks @StrongMonkey

**Comment by Gambler13** [comment]
I finally managed it, thank you guys
For all those who struggeling to set the correct value as answer in Rancher
![image](https://user-images.githubusercontent.com/24626221/41589931-1573fe82-73b6-11e8-964b-5a9ff003c52b.png)

Use the following format
[questionKey].[answerKey] = [answerValue]

**Comment by lmangani** [comment]
@Gambler13 could you share the question format you used to assemble the key-pair array?

**Comment by Gambler13** [comment]
The question format is `controller.nodeSelector = { } ` (see screenshot)

![image](https://user-images.githubusercontent.com/24626221/41600972-825af4e4-73d7-11e8-90f4-924b440d19aa.png)

**Comment by Gambler13** [comment]
My example is for the **dict/hash** 
As selmison mentioned there is another syntax for **array/list** 
I guess you have to use something like `servers={server1, server2}`

**Comment by wmolyneux** [comment]
I am having the same issue as @lmangani 

Did anyone have the question.yaml for an array of answers? Seems like theres been no concrete answer.

Thanks in advance

**Comment by Remigius2011** [comment]
Hi,

I had the same problem when trying to start an instance of a PostgreSQL server from the catalog. To run it e.g. on a node having the tag `mytag=hello`, I used `nodeSelector.mytag` in the Variable field and `hello` in the Value field of the anaswer. It is also possible to paste directly `nodeSelector.mytag=hello` into the Variable field. As I did not find this as an answer anywhere, I post this here in case someone having the same problem drops here.

Cheers, Remi.

**Comment by codejamninja** [comment]
@Gambler13 I don't think you're talking about the same thing. You're example is using a helm template, not a rancher 2 catalog helm template (with the `questions.yaml` file). I think what we're trying to figure out is how to add array or map questions to the `questions.yaml` file.

**Comment by loganhz** [comment]
Hi @lmangani 

Are you talking about how to add array or map questions to the `questions.yaml` file?

If you are talking about how to add array answer, please try the following. It's an example about how to add `tolerations` to kubernetes dashboard from helm stable

You can click edit in YAML (need to use v2.1.0) or input them as key/vale pairs

![image](https://user-images.githubusercontent.com/21168270/46618905-43803580-cb53-11e8-994e-4922bb5b7b19.png)
![image](https://user-images.githubusercontent.com/21168270/46618947-5eeb4080-cb53-11e8-80dd-05351bd71119.png)

**Comment by codejamninja** [comment]
Yeah, but how do you do that when a `questions.yaml` file exists.

**Comment by rstpv** [comment]
In the 2.1 release there is a new feature:

**Comment by loganhz** [comment]
@rstpv thanks for your update. 

It is only a front-end enhancement. It will convert the yaml to key/value pairs in frontend. And pass them to backend.

**Comment by loganhz** [comment]
cc @guangbochen @StrongMonkey How can we support array as a type in `questions.yaml`

**Comment by codejamninja** [comment]
Thank @loganhz. @rstpv, the "feature" you're describing is the issue I'm talking about. I'm completely aware of the rancher questions UI sugar. My problem is it disables modifying helm variables not added to questions.yaml and it does not support dynamic arrays.

**Comment by fabiorauber** [comment]
It's a pity we have to "Edit as YAML" if we need to set node selectors, taints or tolerations. Doing everything through Rancher forms is the feature I miss the most from Rancher 1.6.

**Comment by zbialik** [comment]
I agree this should be something provided by Rancher inherently, especially as a UI feature.

For those that need to provide an override value (answer) for a value that is a list/array type, you can do it likeso:

helm chart values.yaml:
```
key1:
  listKey:
    - item1
    - item2
```

your overriding answer for the above: `key1.listKey="{item1,item2,item3}"`

**NOTE: the most important thing I have found is that this does NOT work if you include spaces after the commas. The above only works if you only have comma delimiter without spaces.**

**Comment by schuerg** [comment]
Is this feature on the roadmap? Adding Rancher Helm Chart install option support via form values for arrays.

**Comment by theAkito** [comment]
Reading the entire conversation, it seems to me that it wasn't clear how "Edit as YAML" is not a solution. It's not even a workaround, because you would break all the advantages of having a `questions.yaml` to begin with, once you hit that button. Therefore, I want to clearly point out, there has to be an `array` type in the `questions.yaml` and we should not be forced to "Edit as YAML" or do some weird alien operations, as proposed in this issue by some.

As for how it should work, I'm sure it's not a big problem to implement it the right way. One way I think of is having a normal question in the `questions.yaml`, just like e.g. a `boolean` type question, appear in the UI, but instead of having a single field to answer, you can add as many fields as you want to the array.

![](https://user-images.githubusercontent.com/21168270/46618947-5eeb4080-cb53-11e8-80dd-05351bd71119.png)

Just like this, but instead of it being a map, you just remove all the keys and keep the values.

![](https://user-images.githubusercontent.com/24626221/41589931-1573fe82-73b6-11e8-964b-5a9ff003c52b.png)

You then just use the plus or minus sign, as much as needed, input the values and then each value will be one value in the array pointed to in the `values.yaml`.

All this is just a matter of stripping down an **already existing feature**, providing it as a type. That's it.

I'm unsure why this issue seems to be on such a low priority, because the whole point of Rancher is to have a UI, which makes managing Kubernetes clusters easier and more comfortable. What's the point of the UI if I have to "Edit as YAML" or do it weirdly, anyway? Might as well just use Kubernetes with Helm only, then.
(Not complaining about a missing feature; I just want to point out how flawful the "suggestions" were, which pointed to just using "Edit as YAML", as if that were some kind of solution or workaround... It's not even a workaround!!)

**Comment by lindhe** [comment]
I'm guessing this relates to https://github.com/rancher/dashboard/issues/5754 too?
